include/nbl/builtin/hlsl/math/functions.hlsl

```diff
@@ -296,50 +313,66 @@ bool partitionRandVariable(float leftProb, NBL_REF_ARG(float) xi, NBL_REF_ARG(fl
 
-template <>
-float conditionalAbsOrMax<float>(bool cond, float x, float limit)
+template <typename T> NBL_PARTIAL_REQ_TOP(concepts::FloatingPointLikeScalar<T>)
+struct conditionalAbsOrMax_helper<T NBL_PARTIAL_REQ_BOT(concepts::FloatingPointLikeScalar<T>) >
 {
-    const float condAbs = asfloat(asuint(x) & uint(cond ? 0x7fFFffFFu : 0xffFFffFFu));
-    return max(condAbs,limit);
-}
+    static T __call(bool cond, NBL_CONST_REF_ARG(T) x, NBL_CONST_REF_ARG(T) limit)
+    {
+        using UintOfTSize = unsigned_integer_of_size_t<sizeof(T)>;
+        const T condAbs = bit_cast<T>(bit_cast<UintOfTSize>(x) & (cond ? 0x7fFFffFFu : 0xffFFffFFu));
 
-template <>
-float32_t2 conditionalAbsOrMax<float32_t2>(bool cond, float32_t2 x, float32_t2 limit)
-{
-    const float32_t2 condAbs = asfloat(asuint(x) & select(cond, (uint32_t2)0x7fFFffFFu, (uint32_t2)0xffFFffFFu));
-    return max(condAbs,limit);
-}
+        return max(condAbs, limit);
+    }
+};
 
-template <>
-float32_t3 conditionalAbsOrMax<float32_t3>(bool cond, float32_t3 x, float32_t3 limit)
+template <typename T> NBL_PARTIAL_REQ_TOP(concepts::FloatingPointLikeVectorial<T>)
+struct conditionalAbsOrMax_helper<T NBL_PARTIAL_REQ_BOT(concepts::FloatingPointLikeVectorial<T>) >
 {
-    const float32_t3 condAbs = asfloat(asuint(x) & select(cond, (uint32_t3)0x7fFFffFFu, (uint32_t3)0xffFFffFFu));
-    return max(condAbs,limit);
-}
+    static T __call(bool cond, NBL_CONST_REF_ARG(T) x, NBL_CONST_REF_ARG(T) limit)
+    {
+        using UintOfTSize = unsigned_integer_of_size_t<sizeof(vector_traits<T>::scalar_type)>;
+        const int dimensionOfT = vector_traits<T>::Dimension;
+        using Uint32VectorWithDimensionOfT = vector<uint32_t, dimensionOfT>;
+        using scalar_type = typename vector_traits<T>::scalar_type;
 
-template <>
-float32_t4 conditionalAbsOrMax<float32_t4>(bool cond, float32_t4 x, float32_t4 limit)
-{
-    const float32_t4 condAbs = asfloat(asuint(x) & select(cond, (uint32_t4)0x7fFFffFFu, (uint32_t4)0xffFFffFFu));
-    return max(condAbs,limit);
-}
-#endif
+        Uint32VectorWithDimensionOfT xAsUintVec;
+        {
+            array_get<T, scalar_type> getter;
+            array_set<Uint32VectorWithDimensionOfT, UintOfTSize> setter;
+
+            for (int i = 0; i < dimensionOfT; ++i)
+                setter(xAsUintVec, i, bit_cast<UintOfTSize>(getter(x, i)));
+        }
+
+        const Uint32VectorWithDimensionOfT mask = cond ? _static_cast<Uint32VectorWithDimensionOfT>(0x7fFFffFFu) : _static_cast<Uint32VectorWithDimensionOfT>(0xffFFffFFu);
+        const Uint32VectorWithDimensionOfT condAbsAsUint = xAsUintVec & mask;
+
+        T condAbs;
+        {
+            array_get<Uint32VectorWithDimensionOfT, UintOfTSize> getter;
+            array_set<T, scalar_type> setter;
+
+            for (int i = 0; i < dimensionOfT; ++i)
+                setter(condAbs, i, bit_cast<scalar_type>(getter(condAbsAsUint, i)));
+        }
```

Comment:
need a separate specialization for native vectors, because you can just `bit_cast` them to their integer counterparts and back

Reply:
looks like there is such specialization implemented already